src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -227,11 +239,44 @@ protected PathFragment readSymbolicLink(PathFragment path) throws IOException {
   @Override
   protected void createSymbolicLink(PathFragment linkPath, PathFragment targetFragment)
       throws IOException {
-    /*
-     * TODO(buchgr): Optimize the case where we are creating a symlink to a remote output. This does
-     * add a non-trivial amount of complications though (as symlinks tend to do).
-     */
-    downloadFileIfRemote(execRoot.getRelative(targetFragment));
+    FileStatus stat = statNullable(linkPath, /* followSymlinks= */ false);
+    if (stat != null) {
+      throw new IOException(linkPath + " (File exists)");
+    }
+
+    // If the output is a regular file or tree artifact, it has "copy" semantics, i.e., it will be
+    // transparently resolved into the artifact it points to; the symlink is just a filesystem-level
+    // optimization to avoid making an actual copy. We can inject its metadata and avoid the need to
+    // compute it from the filesystem later, which would be slower and require additional filesystem
+    // operations to be implemented.
```

Comment:
Potentially, we could download the same files twice. e.g. symlink `a` points to file `b`, but the metadata are same from the perspective of skyframe.

When staging inputs for actions that requires `a`, we download `a` and create files instead of symlinks on the local file system. Then we do the same thing for actions that requires `b`. It is getting worse if the target is directory. This is another reason why we currently creating a dangling symlink for the internal equivalent.

Reply:
Per offline discussion, I've added a `stagingExecPath` field to `RemoteFileArtifactValue`, which is used by the input prefetcher to avoid downloading the same file more than once. This is a bit of a hack, but the alternatives all seem excessively complex to me.

Reply:
Let me write down what we have discussed just now:
- For file, the `stagingExecPath` is used to store the path of final target file. We might want to rename it to a better name to reflect the fact that this remote metadata is actually a symlink to another remote metadata. Alternative way is to create a subclass of `RemoteFileArtifactValue`. I'm fine with either way.
- For tree, we don't want to store `stagingExecPath` in remote metadata for every `TreeFileArtifact`. We can store it in an extra field of `TreeArtifactValue` or in a subclass.
- We resolve symlink chains to the last target when injecting the metadata in the `createSymbolicLink`. So when we staging the symlink in the prefetcher, we know the `execPath` of final target (This is what you are doing now, just write it down for clarification).
- For symlinks whose targets are on the local file system (i.e. remote metadata for the target is `null`), we don't inject metadata but just create symlinks on the local files system.
- We need to update the serialization/deserialization code for remote metadata in https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/actions/cache/CompactPersistentActionCache.java.
- We are interested in the test case: `sym` points to a `dir/file` (by `declare_file`) inside a `dir` (by `declare_dir`)
  - are we able to deduplicate the downloads of `dir/file`?

I might miss some other points but you get the idea.

Reply:
PTAL. Some points of note:

- Naming: I've settled on `symlinkTargetExecPath` instead of `stagingExecPath`. I think that's as explicit as it gets :)
- Skyframe values: I've added a `symlinkTargetExecPath` field to the existing `TreeArtifactValue` (whose size remains at 40 bytes due to padding) and introduced a new`SymlinkToRemoteFileArtifactValue` subclass (to avoid increasing the size of `RemoteFileArtifactValue`). Following a suggestion by Alex, the metadata for tree artifacts is retrieved via `TreeArtifactValue#getMetadata`, which avoids adding a `MetadataProvider#getTreeArtifactValue` method.
- Nested artifacts: I've added tests for them in `BuildWithoutTheBytesIntegrationTest`, but as we discussed, the prefetcher currently isn't able to deduplicate the downloads (and the downloads might even collide). Since this is a preexisting issue and nested artifacts are relatively rare, I'd prefer to fix it outside of this PR. It has been filed as #16333.

Reply:
Totally make sense. Thanks!